tools/src/main/java/org/apache/kafka/tools/GetOffsetShell.java

```diff
@@ -0,0 +1,404 @@
+            } catch (OptionException e) {
+                throw new TerseException(e.getMessage());
+            }
+
+            if (options.has(bootstrapServerOpt)) {
+                effectiveBrokerListOpt = bootstrapServerOpt;
+            } else {
+                effectiveBrokerListOpt = brokerListOpt;
+            }
+
+            CommandLineUtils.checkRequiredArgs(parser, options, effectiveBrokerListOpt);
+
+            String brokerList = options.valueOf(effectiveBrokerListOpt);
+
+            try {
+                ToolsUtils.validateBootstrapServer(brokerList);
+            } catch (IllegalArgumentException e) {
+                CommandLineUtils.printUsageAndExit(parser, e.getMessage());
+            }
+        }
+
+        public boolean hasTopicPartitionsOpt() {
+            return options.has(topicPartitionsOpt);
+        }
+
+        public String topicPartitionsOpt() {
+            return options.valueOf(topicPartitionsOpt);
+        }
+
+        public boolean hasTopicOpt() {
+            return options.has(topicOpt);
+        }
+
+        public String topicOpt() {
+            return options.valueOf(topicOpt);
+        }
+
+        public boolean hasPartitionsOpt() {
+            return options.has(partitionsOpt);
+        }
+
+        public String partitionsOpt() {
+            return options.valueOf(partitionsOpt);
+        }
+
+        public String timeOpt() {
+            return options.valueOf(timeOpt);
+        }
+
+        public boolean hasCommandConfigOpt() {
+            return options.has(commandConfigOpt);
+        }
+
+        public String commandConfigOpt() {
+            return options.valueOf(commandConfigOpt);
+        }
+
+        public String effectiveBrokerListOpt() {
+            return options.valueOf(effectiveBrokerListOpt);
+        }
+
+        public boolean hasExcludeInternalTopicsOpt() {
+            return options.has(excludeInternalTopicsOpt);
+        }
+    }
+
+    public Map<TopicPartition, Long> fetchOffsets(GetOffsetShellOptions options) throws IOException, ExecutionException, InterruptedException, TerseException {
+        String clientId = "GetOffsetShell";
+        String brokerList = options.effectiveBrokerListOpt();
+
+        if (options.hasTopicPartitionsOpt() && (options.hasTopicOpt() || options.hasPartitionsOpt())) {
+            throw new TerseException("--topic-partitions cannot be used with --topic or --partitions");
+        }
+
+        boolean excludeInternalTopics = options.hasExcludeInternalTopicsOpt();
+        OffsetSpec offsetSpec = parseOffsetSpec(options.timeOpt());
+
+        TopicPartitionFilter topicPartitionFilter;
+
+        if (options.hasTopicPartitionsOpt()) {
+            topicPartitionFilter = createTopicPartitionFilterWithPatternList(options.topicPartitionsOpt());
+        } else {
+            topicPartitionFilter = createTopicPartitionFilterWithTopicAndPartitionPattern(
+                    options.hasTopicOpt() ? options.topicOpt() : null,
```

Comment:
I think `options.topicOpt()` returns `null` if it's unset, so maybe we don't need to check `options.hasTopicOpt()`?